---
title: "GPR68 Deep Mutational Scan Multiphenotypic Analysis"
author: "Matthew Howard, UCSF"
date: "March 11, 2024"
output: html_notebook
---
this notebook contains code used to analyze multiphenotypic GPR68 deep mutational scanning data and prepare plots

screening conditions: pH 6.5, pH 5.5, and surface expression

abbreviations and general naming conventions:
decimals are omitted when referring to pH conditions, example:"ph55" for pH 5.5

set working directory, reset to source folder on your system 
```{r}
setwd("/Users/mkh/Library/CloudStorage/Box-Box/WCM Lab/Matt/Lab/Projects/proton_sensors/GPR68_DMS/figures") 
```

source R packages
```{r}
install.packages("tidyverse")
install.packages("ggsci")
install.packages("ggpubr")
install.packages("gridExtra")
install.packages("cowplot")
install.packages("bio3d")
install.packages("tune")
install.packages("MASS")
install.packages("colorspace")
insall.packages("patchwork")

library(tidyverse)
library(ggsci)
library(ggpubr)
library(gridExtra)
library(cowplot)
library(bio3d)
library(tune)
library(MASS)
library(colorspace)
library(patchwork)
```

pull in DMS data and gpr68 structure
```{r}
gpr68_scores<-read.csv('gpr68_scores_processed.csv')

colnames(gpr68_scores) <- c("hgvs", "ph55_SE", "ph55_epsilon", "ph55_score", "ph65_SE", "ph65_epsilon", "ph65_score", "surf_SE", "surf_epsilon", "surf_score",'pos','len','mutation_type','variants','is.wt')

order_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")
gpr68_scores_missense <- gpr68_scores %>% filter(variants %in% order_missense)

gpr68_structure <- read.pdb("structures/gpr68_gsq_structure.pdb")
```

define structure map function
borrowed from Christian Macdonald's (@UCSF) work as part of https://github.com/odcambc/DIMPLE_manuscript_figures
```{r}
# Function to map scores to a PDB structure based on residue numbers
# Input:
#   input_pdb: Input PDB structure
#   mapping_scores: Dataframe containing mapping scores
#   field: The field in the mapping_scores dataframe to use for scoring
#   selection: Optional selection of atoms, default is "protein"

map_scores_pdb <- function(input_pdb, mapping_scores, field, selection = NULL) {
  
  # If selection is not provided, select all atoms in the "protein" segment
  if (is.null(selection)) {
    selection = atom.select(input_pdb, "protein")
  }
  
  # Trim the PDB structure to the selected atoms
  output_pdb = trim.pdb(input_pdb, selection)
  
  # Loop through each atom in the trimmed PDB structure
  for (i in seq_len(dim(output_pdb$atom)[1])) {
    
    # Check if the residue number is greater than 0
    if (output_pdb$atom[i,]$resno > 0) {
      
      # Convert the residue number to a character
      n = as.character(output_pdb$atom[i,]$resno)
      
      # Find the corresponding row in mapping_scores where 'pos' matches the residue number
      j = which(mapping_scores['pos'] == n)
      
      # Check if there is a matching row in mapping_scores
      if (length(j) == 0) {
        score = 0
      } else {
        # Get the score from the specified field in the matching row
        score = mapping_scores[j, field][[1]]
      }
      
      # Check if the score is not NA (not missing)
      if (!is.na(score)) {
        # Update the 'b' column in the PDB structure with the score
        output_pdb$atom[i,]$b = score
      } else {
        # If the score is missing, set 'b' to 0
        output_pdb$atom[i,]$b = 0
      }
    } else {
      # If the residue number is not greater than 0, set 'b' to 0
      output_pdb$atom[i,]$b = 0
    }
  }
  
  # Return the PDB structure with updated scores
  return(output_pdb)
}
```

make scatter of expression vs function scores, fit line to syn population, calculate euclidean distance from line for all variants, above line are GOF below line are LOF
make waterfall plot averaging distances/position for positive (GOF) and negative (LOF) distances
```{r}
# custom theme
  theme_options <- theme(
    panel.border = element_rect(color = "black",fill = NA, size = 1),
    text = element_text(color = "black", family = "sans", size = 10),
    legend.position = "right",
    axis.line = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 10, family = "sans", color = "black"),
    axis.text = element_text(size = 10, family = "sans", color = "black"),
    legend.text = element_text(size = 10, family = "sans", color = "black"),
    legend.title = element_text(size = 10, family = "sans", color = "black"),
    plot.subtitle = element_text(size = 10, family = "sans", color = "black"),
    plot.caption = element_text(size = 10, family = "sans", color = "black"),
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank()
  )

# structural and sequence pos of interest
  gaq_contacts <- c(48, 54, 55, 56, 60, 122, 123, 126, 127, 130, 133, 205, 208, 211, 212, 215, 216, 217, 218, 219, 222, 226, 229, 231, 287, 289, 290, 291) #excludes those in other motifs
  DPxxY <- c(282, 283, 284, 285, 286)
  DRY <- c(118, 119, 120)
  CFxP <- c(240, 241, 242, 243)
  PIF <- c(109, 197, 237)

  # ecl <- c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17, 18, 19, 20, 21, 22,75, 76, 77, 78, 79, 80, 81, 82, 83,84,85,86,87,88,89,90,91,92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 148, 149, 150, 151, 152, 153, 154, 155,   156, 157, 158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266,   267, 268, 269, 271)

# Define a function for creating a scatter plot and waterfalls based on euclidean distance metric
scatter_waterfalls <- function(data, x_var, y_var, y_var_for_line, output_file1, output_file2) {
  # labels for x and y var
  x_var_label <- ifelse(x_var == "surf_score", "Surface Expression", x_var)
  y_var_label <- ifelse(y_var == "surf_score", "Surface Expression",
                        ifelse(y_var == "ph55_score", "pH 5.5",
                               ifelse(y_var == "ph65_score", "pH 6.5", y_var)))
  
  # fit a linear regression model to the data using syn mutants of y_var_for_line and surf_score
  lm_model <- lm(data %>% filter(is.wt == TRUE) %>% dplyr::select(y_var_for_line, surf_score))
  lm_model_all <- lm(data %>% dplyr::select(y_var_for_line, surf_score))
  
  # obtain the coefficients (a, b, c) from the lm_model
  a <- coef(lm_model)[2]
  b <- -1  # Assuming b as -1 for perpendicular distance
  c <- coef(lm_model)[1]
 
  # calculate the R-squared value from the linear regression model
  r_squared <- summary(lm_model)$r.squared
  r_squared_all <- summary(lm_model_all)$r.squared
  
  # Format the R-squared value as a string
  r_squared_label <- paste("R-squared(syn/all) = ", round(r_squared, 2), "/", round(r_squared_all, 2))

  # Calculate the signed distances from each point (x_var, y_var) to the line
  data$distance <- -(a * data[[x_var]] + b * data[[y_var]] + c) / sqrt(a^2 + b^2) 

  # Create a dataFrame for distances
  distance_df <- data.frame(hgvs = data$hgvs, surf_score = data$surf_score, ph55_score = data$ph55_score, ph65_score = data$ph65_score, distance = data$distance, pos = data$pos, is.wt = data$is.wt, variants = data$variants)
  # Remove rows with missing distances in the distance_df
  distance_df <- distance_df %>% filter(!is.na(distance))
  
  # Identify the top 5% highest distances
  top_5_percent_cutoff <- quantile(distance_df$distance, probs = 0.975)
  top_5_percent_data <- distance_df %>% filter(distance >= top_5_percent_cutoff)
  bottom_5_percent_cutoff <- quantile(distance_df$distance, probs = 0.025)
  bottom_5_percent_data <- distance_df %>% filter(distance <= bottom_5_percent_cutoff)

  # Add a "percentile" column to the input dataframe
  data$percentile <- ifelse(data$hgvs %in% top_5_percent_data$hgvs, "top", ifelse(data$hgvs %in% bottom_5_percent_data$hgvs, "bottom", "neutral"))

  # Separate distances into two data frames: below and above the best fit line
  below_df <- distance_df[distance_df$distance < 0, ]
  above_df <- distance_df[distance_df$distance >= 0, ]

  # Calculate the absolute sum of distances by position and divide by the count
  above_mean_pos_df <- above_df %>%
    group_by(pos) %>%
    summarize(mean_distance = mean(distance))
  
  below_mean_pos_df <- below_df %>%
    group_by(pos) %>%
    summarize(mean_distance = mean(distance))

  
  
  
  # Create a scatter plot with grey color for all points and blue for "top" points
  scatter_plot <- ggscatter(distance_df, x = x_var, y = y_var, size = 1, shape = 21, fill = "distance", stroke = 0) +
    labs(x = x_var_label, y = y_var_label) +
    theme_options + 
    ggtitle(paste0(x_var_label, " vs ", y_var_label, " (mis/syn dataset)\n", r_squared_label )) +
    geom_smooth(data = data %>% filter(is.wt == TRUE), aes(x = surf_score, y = !!sym(y_var_for_line)), method = "lm", color = "black", size = 0.5, fill = "black", alpha = 0.2, fullrange = TRUE) +
    scale_fill_gradient2(low = "red", mid = "white", high = "blue", midpoint = 0) +
    geom_point(data = data %>% filter(is.wt == TRUE), aes(x = surf_score, y = !!sym(y_var_for_line)), size = 0.25, alpha = 0.75, color = "black") +
    theme(legend.position = "none")

  # create a new column 'group' in the 'above_sum_df' data frame based on the 'pos' values
  above_mean_pos_df$group <- NA  # point fill
  above_mean_pos_df$border <- "none"  # point border

  above_mean_pos_df$group[above_mean_pos_df$pos %in% gaq_contacts] <- "gaq"
  above_mean_pos_df$group[above_mean_pos_df$pos %in% DPxxY] <- "DPxxY"
  above_mean_pos_df$group[above_mean_pos_df$pos %in% DRY] <- "DRY"
  above_mean_pos_df$group[above_mean_pos_df$pos %in% CFxP] <- "CFxP"
  above_mean_pos_df$group[above_mean_pos_df$pos %in% PIF] <- "PIF"
  # above_sum_df$border[above_sum_df$pos %in% ecl] <- "ecl"

  # create a new column 'group' in the 'above_sum_df' data frame based on the 'pos' values
  below_mean_pos_df$group <- NA  # point fill
  below_mean_pos_df$border <- "none"  # point color

  below_mean_pos_df$group[below_mean_pos_df$pos %in% gaq_contacts] <- "gaq"
  below_mean_pos_df$group[below_mean_pos_df$pos %in% DPxxY] <- "DPxxY"
  below_mean_pos_df$group[below_mean_pos_df$pos %in% DRY] <- "DRY"
  below_mean_pos_df$group[below_mean_pos_df$pos %in% CFxP] <- "CFxP"
  below_mean_pos_df$group[below_mean_pos_df$pos %in% PIF] <- "PIF"
  # below_sum_df$border[below_sum_df$pos %in% ecl] <- "ecl"

  above_mean_pos_df$pos <- as.numeric(above_mean_pos_df$pos)  # Convert 'pos' to numeric
  below_mean_pos_df$pos <- as.numeric(below_mean_pos_df$pos)  # Convert 'pos' to numeric

  distance_plot_above <- ggplot(data = above_mean_pos_df %>%
         arrange(mean_distance), aes(x = log(as.numeric(reorder(pos, desc(mean_distance)))), y = mean_distance)) +
    labs(x = "Rank", y = "Mean Pos. Distance") +
    geom_point(aes(fill = group), size = 3, shape = 21) +
    scale_x_continuous(labels = c(1, 10, 100), 
                       breaks = log(c(1, 10, 100)), 
                       expand = c(.025, .025)) +
    ggtitle(y_var_label) +
    theme_options +
    theme(panel.background = element_rect(fill = "white"),
          axis.title = element_blank()) +
    scale_y_continuous(limits = c(0, max(above_mean_pos_df$mean_distance)), expand = c(0.07,0)) +
    scale_color_manual(values = c("none" = rgb(0, 0, 0, alpha=0), "ecl" = rgb(0, 0, 0, alpha=0)))

  # Create a plot for ranked positions and their absolute sum of distances
  distance_plot_below <- ggplot(data = below_mean_pos_df %>%
         arrange(mean_distance), aes(x = log(as.numeric(reorder(pos, mean_distance))), y = mean_distance)) +
 
    labs(x = "Rank, log-scale", y = "Mean Pos. Distance") +
    geom_point(aes(fill = group), size = 3, shape = 21) +
    scale_x_continuous(labels = c(1, 10, 100), 
                       breaks = log(c(1, 10, 100)), 
                       expand = c(.025, .025)) +
    theme_options +
    theme(panel.background = element_rect(fill = "white"),
        axis.title = element_blank()) +
    scale_y_continuous(limits = c(min(below_mean_pos_df$mean_distance), expand = c(0, 0))) +
      scale_color_manual(values = c("none" = rgb(0, 0, 0, alpha=0), "ecl" = rgb(0, 0, 0, alpha=0)))

  # patchwork arrange distance plot vertically
  plot_arranged <- distance_plot_above / distance_plot_below
    
  # Save the plot as an output file
  ggsave(output_file1, scatter_plot, height = 80, width = 80, units = "mm",  dpi = 450)
  ggsave(output_file2, plot_arranged, height = 130, width = 140, units = "mm", dpi = 450)
  
  # Return the distance df
  return(list(above_mean_pos_df, below_mean_pos_df, distance_df))
}

# generate plots, use only gpr68_scores_missense df
ph55_distance <- scatter_waterfalls(gpr68_scores_missense, "surf_score", "ph55_score",y_var_for_line = "ph55_score", "main/figure4/GPR68_surf_ph55_scatter.pdf", "supplemental/supp_figure17/GPR68_surf_ph55_ranked_pos.pdf")

ph65_distance <- scatter_waterfalls(gpr68_scores_missense, "surf_score", "ph65_score",y_var_for_line = "ph65_score", "main/figure4/GPR68_surf_ph65_scatter.pdf", "supplemental/supp_figure17/GPR68_surf_ph65_ranked_pos.pdf")
```

map GOF/LOF scores to structure
```{r}
#data for export elsewhere 
ph55_LOF <- ph55_distance[[2]]
write.table(ph55_LOF, file='ph55_LOF.tsv', quote=FALSE, sep='\t')
ph55_GOF <- ph55_distance[[1]]
write.table(ph55_GOF, file='ph55_GOF.tsv', quote=FALSE, sep='\t')
ph65_LOF <- ph65_distance[[2]]
write.table(ph55_LOF, file='ph65_LOF.tsv', quote=FALSE, sep='\t')
ph65_GOF <- ph65_distance[[1]]
write.table(ph55_GOF, file='ph65_GOF.tsv', quote=FALSE, sep='\t')

x = map_scores_pdb(gpr68_structure, ph55_distance[[1]], "mean_distance")
ca.chainsABC<-atom.select(x, 'calpha', chain = c('R'))
write.pdb(x, file="main/figure5/GPR68_ph55_GOF.pdb")

x = map_scores_pdb(gpr68_structure, ph55_distance[[2]], "mean_distance")
ca.chainsABC<-atom.select(x, 'calpha', chain = c('R'))
write.pdb(x, file="main/figure5/GPR68_ph55_LOF.pdb")

x = map_scores_pdb(gpr68_structure, ph65_distance[[1]], "mean_distance")
ca.chainsABC<-atom.select(x, 'calpha', chain = c('R'))
write.pdb(x, file="main/figure5/GPR68_ph65_GOF.pdb")

x = map_scores_pdb(gpr68_structure, ph65_distance[[2]], "mean_distance")
ca.chainsABC<-atom.select(x, 'calpha', chain = c('R'))
write.pdb(x, file="main/figure5/GPR68_ph65_LOF.pdb")
```

make boxplot of all surface-adusted ph55 scores and then subset of the ones around the pocket

simply make a group column, pocket 1, 2, rest and then plot the violins w/barplot on top using ggviolin 
define pocket 1, 2, both, rest
pocket 1:13, 14, 15, 17,20,28,77,80,85,101,102,164,165,166,167,169,171,172,174,175,176,265,269,272,273
pocket 2: 102, 156, 162, 173, 174, 178, 181, 182, 186, 189, 244, 248, 251, 268

```{r}
ph55_distance_scores <- ph55_distance[[3]]

ph55_distance_scores$group <- "non-pocket"
#define groups, remove 13, 14, 165, 166, 167 from pocket 1 as the side chains point into extracellular space
ph55_distance_scores$group <- ifelse(
    ph55_distance_scores$pos %in% c(15, 17,20,28,77,80,85,101,102,164,169,171,172,174,175,176,265,269,272,273),
    "pocket",ph55_distance_scores$group)

violins <- ggplot(ph55_distance_scores, aes(x = group, y = distance, color = group)) + 
  geom_hline(yintercept = 0, linetype = "dotted", color = "black", size = 0.3) +
  geom_violin(size = 0.3) +
  geom_boxplot(width = 0.2, color = "black", outlier.size = 0.3, size = 0.5) +
  theme_minimal() +
  labs(x = "",
       y = "Expression-adjusted signaling score") +
  scale_color_manual(values = c("non-pocket" = "black", "pocket" = "red")) +  # Set fill colors for each group
  theme(legend.position = "none",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line = element_line(color = "black", size = 0.25),
        text = element_text(size = 6, family = "sans"),
        axis.text = element_text(color = "black"),
        axis.ticks = element_line(color = "black", size = 0.25),
        axis.ticks.length = unit(0.05, "cm")) +
  scale_y_continuous(breaks = c(-2.5, 0, 2.5))

violins

ggsave("main/figure5/ph55_violins.pdf", violins, width = 30, height = 55, units = "mm", dpi = 600)

# stats testing

# Subset the data for each group
nonpocket_data <- ph55_distance_scores$distance[ph55_distance_scores$group == "non-pocket"]
pocket_data <- ph55_distance_scores$distance[ph55_distance_scores$group == "pocket"]

# Perform t-test
t_test_result <- t.test(nonpocket_data, pocket_data)

# Print the result
print(t_test_result)

# Perform Mann-Whitney U test
wilcox_test_result <- wilcox.test(nonpocket_data, pocket_data)

# Print the result
print(wilcox_test_result)
```

histidine heatmap snapshots, could be better code but it works
```{r}
# filter the ph55_distance_sum_signed[[3]] data frame to histidine residues
ph55_his <- ph55_distance_sum_signed[[3]] %>% filter(pos %in% c(17,20,84,89,159,169,175,245,269))

# in the pos column of ph55_his change each occurrance of 17 to 1, 20 to 2, 84 to 3, 89 to 4, 159 to 5, 169 to 6, 175 to 7, 245 to 8, 269 to 9
ph55_his$pos <- recode(ph55_his$pos, "17" = "2", "20" = "3", "84" = "4", "89" = "5", "159" = "6", "169" = "7", "175" = "8", "245" = "9", "269" = "10")

order_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")
label_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")
ph55_his$distance <- as.numeric(ph55_his$distance)
ph55_his$pos <- as.numeric(ph55_his$pos)

# add a row to ph55_his for each position, except pos = 8, and set the variatns to H
ph55_his <- ph55_his %>% add_row(pos = 1, variants = "H") %>% add_row(pos = 2, variants = "H") %>% add_row(pos = 3, variants = "H") %>% add_row(pos = 4, variants = "H") %>% add_row(pos = 5, variants = "H") %>% add_row(pos = 6, variants = "H") %>% add_row(pos = 7, variants = "H") %>% add_row(pos = 9, variants = "H") %>% add_row(pos = 10, variants = "H")

# add a row to ph55_his for pos = 8, set the distane to 0 and set the variatns to T
ph55_his <- ph55_his %>% add_row(pos = 8, variants = "T", distance = NA)
ph55_his <- ph55_his %>% add_row(pos = 8, variants = "C", distance = NA)

create_heatmap <- function(gene, order, chunk_size, scores_df, score_column, variant_names = NULL, output_file = NULL) {
  
  # Check if variant_names were provided; if not, use the original variant names
  if (is.null(variant_names)) {
    variant_names <- order
  }
  
  scores_df <- scores_df %>%
    filter(variants %in% order)
  
  gene_chunks <- str_split(gene, '')[[1]]
  num_chunks <- ceiling(length(gene_chunks) / chunk_size)

  min_score <- min(scores_df[, score_column], na.rm = TRUE)
  max_score <- max(scores_df[, score_column], na.rm = TRUE)

  heatmap_plots <- list()

  for (i in 1:num_chunks) {
    start_idx <- (i - 1) * chunk_size + 1
    end_idx <- min(i * chunk_size, length(gene_chunks))

    gene_chunk <- gene_chunks[start_idx:end_idx]
    scores_chunk <- scores_df[scores_df$pos %in% c(start_idx:(start_idx + length(gene_chunk) - 1)), ]
    
    # Replace the variant names with the desired names
    scores_chunk$variants <- factor(scores_chunk$variants, levels = order, labels = variant_names)

    
    heatmap_plot <- ggplot(data = scores_chunk, aes(x = pos, y = factor(variants, level = variant_names), fill = !!sym(score_column))) +
      geom_tile(aes(color = "grey"), linewidth = 0.2, position = "identity") +
      scale_fill_continuous_divergingx(palette = 'RdBu', mid = 0, l1 = 0.2, l3 = 0.2, p1 = 0.9, p3 = .4, p4 = 0.7, rev = FALSE, na.value = 'grey', limits = c(min_score, max_score)) +
theme(
        panel.background = element_rect(fill = "grey", size = 0.1, linetype = "solid"),
        panel.grid.major = element_line(size = 0.3, linetype = 'solid', colour = "grey"),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_blank(),
        axis.ticks.x.top = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(family = "sans", size = 9, angle = 0, hjust = 0.5, vjust = 1, margin = margin(t = 1)),
        axis.text = element_text(family = "sans", size = 9, color = "black"),
        axis.text.x.top = element_text(family = "sans",  size = 9, angle = 0, hjust = 0.5, margin = margin(r = 0)),
        axis.text.y = element_text(family = "sans",  size = 9,  margin = margin(r = 0)) ) +
      scale_x_continuous(breaks = seq(0, nchar(gene), by = 5), expand = c(0, 0),
                         sec.axis = sec_axis(trans = ~., name = "Sequence", breaks = seq(start_idx, end_idx), labels = gene_chunk, guide = derive())) +
      coord_fixed(ratio = 1) +
      theme(axis.text.y = element_blank(),
            axis.title.y = element_blank()) +
      scale_color_manual(values = c(NA, 'grey'), guide = FALSE) +
      labs(x = "Position") +
      
      geom_tile(data = subset(scores_chunk, !is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE) +
      geom_tile(data = subset(scores_chunk, is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE)
    
    heatmap_plots[[i]] <- heatmap_plot
  }
  #heatmap_combined <- ggarrange(plotlist = heatmap_plots, nrow = num_chunks, ncol = 1)
  
  # Assuming heatmap_plots is your list of plots
    subset_plots <- heatmap_plots[2:10]

    # Use patchwork::wrap_plots with the subset_plots
    heatmap_combined <- patchwork::wrap_plots(subset_plots, nrow = 1, ncol = length(subset_plots)) + 
  plot_layout(guides = 'collect') 


  #heatmap_combined <- patchwork::wrap_plots(heatmap_plots, nrow = 1, ncol = num_chunks)
    ggsave(output_file, heatmap_combined, height = 20, width = 10, dpi = 450)
}

create_heatmap(gene = histidine_seq, order = order_missense, chunk_size = 1, scores_df = ph55_his, score_column = "distance", variant_names = label_missense, output_file = "main/figure5/histidine_heatmap.pdf")
```

DRY&DPxxY heatmap snapshots, could also be more elegant. 
```{r}
# filter the ph55_distance_sum_signed[[3]] data frame to DRY and DPxxY motifs
order_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")
label_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")

dpxxy <- ph55_distance_sum_signed[[3]] %>% filter(pos %in% c(282,283,284,285,286))
dpxxy$pos <- recode(dpxxy$pos,"282" = "1", "283" = "2", "284" = "3", "285" = "4", "286" = "5")
dpxxy$distance <- as.numeric(dpxxy$distance)
dpxxy$pos <- as.numeric(dpxxy$pos)

dry <- ph55_distance_sum_signed[[3]] %>% filter(pos %in% c(118,119,120))
dry$pos <- recode(dry$pos, "118" = "1", "119" = "2", "120" = "3")
dry$distance <- as.numeric(dry$distance)
dry$pos <- as.numeric(dry$pos)

create_heatmap <- function(gene, order, chunk_size, scores_df, score_column, variant_names = NULL, output_file = NULL) {
  
  # Check if variant_names were provided; if not, use the original variant names
  if (is.null(variant_names)) {
    variant_names <- order
  }
  
  scores_df <- scores_df %>%
    filter(variants %in% order)
  
  gene_chunks <- str_split(gene, '')[[1]]
  num_chunks <- ceiling(length(gene_chunks) / chunk_size)

  min_score <- min(scores_df[, score_column], na.rm = TRUE)
  max_score <- max(scores_df[, score_column], na.rm = TRUE)

  heatmap_plots <- list()

  for (i in 1:num_chunks) {
    start_idx <- (i - 1) * chunk_size + 1
    end_idx <- min(i * chunk_size, length(gene_chunks))

    gene_chunk <- gene_chunks[start_idx:end_idx]
    scores_chunk <- scores_df[scores_df$pos %in% c(start_idx:(start_idx + length(gene_chunk) - 1)), ]
    
    # Replace the variant names with the desired names
    scores_chunk$variants <- factor(scores_chunk$variants, levels = order, labels = variant_names)
    
    heatmap_plot <- ggplot(data = scores_chunk, aes(x = pos, y = factor(variants, level = variant_names), fill = !!sym(score_column))) +
      geom_tile(aes(color = "grey"), linewidth = 0.2, position = "identity") +
      scale_fill_continuous_divergingx(palette = 'RdBu', mid = 0, l1 = 0.2, l3 = 0.2, p1 = 0.9, p3 = .4, p4 = 0.7, rev = FALSE, na.value = 'grey', limits = c(min_score, max_score)) +
theme(
        panel.background = element_rect(fill = "grey", size = 0.1, linetype = "solid"),
        panel.grid.major = element_line(size = 0.3, linetype = 'solid', colour = "grey"),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_blank(),
        axis.ticks.x.top = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(family = "sans", size = 9, angle = 0, hjust = 0.5, vjust = 1, margin = margin(t = 1)),
        axis.text = element_text(family = "sans", size = 9, color = "black"),
        axis.text.x.top = element_text(family = "sans",  size = 9, angle = 0, hjust = 0.5, margin = margin(r = 0)),
        axis.text.y = element_text(family = "sans",  size = 9,  margin = margin(r = 0)) ) +
      scale_x_continuous(breaks = seq(0, nchar(gene), by = 5), expand = c(0, 0),
                         sec.axis = sec_axis(trans = ~., name = "Sequence", breaks = seq(start_idx, end_idx), labels = gene_chunk, guide = derive())) +
      coord_fixed(ratio = 1) +
      theme(axis.text.y = element_blank(),
            axis.title.y = element_blank()) +
      scale_color_manual(values = c(NA, 'grey'), guide = FALSE) +
      labs(x = "Position") +
      
      geom_tile(data = subset(scores_chunk, !is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE) +
      geom_tile(data = subset(scores_chunk, is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE)
    
    heatmap_plots[[i]] <- heatmap_plot
  }
  heatmap_combined <- ggarrange(plotlist = heatmap_plots, nrow = num_chunks, ncol = 1)

  heatmap_combined <- patchwork::wrap_plots(heatmap_plots, nrow = 1, ncol = num_chunks)
    ggsave(output_file, heatmap_combined, height = 20, width = 10, dpi = 450)
}

dry_seq <- "DRY"
dpxxy_seq <- "DPILY"

create_heatmap(gene = dry_seq, order = order_missense, chunk_size = 10, scores_df = dry, score_column = "distance", variant_names = label_missense, output_file = "supplemental/supp_figure18/dry_heatmap.pdf")

create_heatmap(gene = dpxxy_seq, order = order_missense, chunk_size = 10, scores_df = dpxxy, score_column = "distance", variant_names = label_missense, output_file = "supplemental/supp_figure18/dpxxy_heatmap.pdf")
```

heatmap of the ph55 surf-adjusted scores
```{r}
order_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")
label_missense <- c("A", "G", "M", "V", "L", "I", "T", "S", "C", "Q", "N", "Y", "W", "F", "E", "D", "H", "K", "R", "P")

gpr68_wt = "MGNITADNSSMSCTIDHTIHQTLAPVVYVTVLVVGFPANCLSLYFGYLQIKARNELGVYLCNLTVADLFYICSLPFWLQYVLQHDNWSHGDLSCQVCGILLYENIYISVGFLCCISVDRYLAVAHPFRFHQFRTLKAAVGVSVVIWAKELLTSIYFLMHEEVIEDENQHRVCFEHYPIQAWQRAINYYRFLVGFLFPICLLLASYQGILRAVRRSHGTQKSRKDQIQRLVLSTVVIFLACFLPYHVLLLVRSVWEASCDFAKGVFNAYHFSLLLTSFNCVADPVLYCFVSETTHRDLARLRGACLAFLTCSRTGRAREAYPLGAPEASGKSGAQGEEPELLTKLHPAFQTPNSPGSGGFPTGRLA"

create_heatmap <- function(gene, order, chunk_size, scores_df, score_column, variant_names = NULL, output_file = NULL) {
  
  # Check if variant_names were provided; if not, use the original variant names
  if (is.null(variant_names)) {
    variant_names <- order
  }
  
  scores_df <- scores_df %>%
    filter(variants %in% order)
  
  gene_chunks <- str_split(gene, '')[[1]]
  num_chunks <- ceiling(length(gene_chunks) / chunk_size)

  min_score <- min(scores_df[, score_column], na.rm = TRUE)
  max_score <- max(scores_df[, score_column], na.rm = TRUE)

  heatmap_plots <- list()

  for (i in 1:num_chunks) {
    start_idx <- (i - 1) * chunk_size + 1
    end_idx <- min(i * chunk_size, length(gene_chunks))

    gene_chunk <- gene_chunks[start_idx:end_idx]
    scores_chunk <- scores_df[scores_df$pos %in% c(start_idx:(start_idx + length(gene_chunk) - 1)), ]
    
    # Replace the variant names with the desired names
    scores_chunk$variants <- factor(scores_chunk$variants, levels = order, labels = variant_names)

    
    heatmap_plot <- ggplot(data = scores_chunk, aes(x = pos, y = factor(variants, level = variant_names), fill = !!sym(score_column))) +
      geom_tile(aes(color = "grey"), linewidth = 0.2, position = "identity") +
      colorspace::scale_fill_continuous_divergingx(palette = 'RdBu', mid = 0, l1 = 0.2, l3 = 0.2, p1 = 0.9, p3 = .4, p4 = 0.7, rev = FALSE, na.value = 'grey', limits = c(min_score, max_score)) +
theme(
        panel.background = element_rect(fill = "grey", size = 0.1, linetype = "solid"),
        panel.grid.major = element_line(size = 0.1, linetype = 'solid', colour = "grey"),
        panel.grid.minor = element_blank(),
        #axis.ticks.x = element_blank(),
        axis.ticks.x.top = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(family = "sans", size = 9, angle = 0, hjust = 0.5, vjust = 1, margin = margin(t = 1)),
        axis.text = element_text(family = "sans", size = 9, color = "black"),
        axis.text.x.top = element_text(family = "sans",  size = 9, angle = 0, hjust = 0.5, margin = margin(r = 0)),
        axis.text.y = element_text(family = "sans",  size = 9,  margin = margin(r = 0)) ) +
      scale_x_continuous(breaks = seq(0, nchar(gene), by = 5), expand = c(0, 0),
                         sec.axis = sec_axis(trans = ~., name = "Sequence", breaks = seq(start_idx, end_idx), labels = gene_chunk, guide = derive())) +
      coord_fixed(ratio = 1) +
      scale_color_manual(values = c(NA, 'grey'), guide = FALSE) +
      labs(y = "Mutation", x = "Position") +
      geom_tile(data = subset(scores_chunk, !is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE) +
      geom_tile(data = subset(scores_chunk, is.wt), aes(color = "grey"), size = 0.2, position = "identity", show.legend = FALSE)
    
    heatmap_plots[[i]] <- heatmap_plot
  }
  heatmap_combined <- ggarrange(plotlist = heatmap_plots, nrow = num_chunks, ncol = 1)
  ggsave(output_file, heatmap_combined, height = 10, width = 20, dpi = 450)
}

create_heatmap(gene = gpr68_wt, order = order_missense, chunk_size = 125, scores_df = ph55_distance_scores, score_column = "distance", variant_names = label_missense, output_file = "main/figure4/ph55_surf_adjusted_heatmap.pdf")

```

